service/src/LevelZero.cpp

```diff
@@ -248,22 +249,33 @@ namespace geopm
                 if (property.onSubdevice == 0) {
                     m_devices.at(device_idx).power_domain = handle;
                     ++num_device_power_domain;
+                    if (num_device_power_domain != 1) {
+                        throw Exception("LevelZero::" + std::string(__func__) +
+                                        ": Multiple device level power domains "
+                                        "detected.  This may lead to incorrect power readings",
+                                        GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                    }
                 }
-#ifdef GEOPM_DEBUG
                 else {
-                    //For initial GEOPM support we're only providing device level power
-                    std::cerr << "Warning: <geopm> LevelZero: A sub-device "
-                                 "level power domain was found but is not currently supported.\n";
+                    ++num_subdevice_power_domain;
+                    m_devices.at(device_idx).subdevice.power_domain.push_back(handle);
                 }
-#endif
+            }
 
-                if (num_device_power_domain != 1) {
-                    throw Exception("LevelZero::" + std::string(__func__) +
-                                    ": GEOPM requires a single device level power domain, "
-                                    "but found " + std::to_string(num_device_power_domain),
-                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
-                }
+            if (num_device_power_domain != 1 && num_subdevice_power_domain == 0) {
```

Comment:
If a system has multiple device power domains and at least one subdevice power domain somewhere, this will not throw.  Could a system be configured that way?  If yes, this needs to be more robust.

Reply:
There is a prior check ~line 252 for:
 if (num_device_power_domain != 1) {          
That should prevent multiple device level power domains from being tracked.  We could consolidate these various checks if needed.

Reply:
Consider if the ```power_domain``` vector had the following entries:
```
    power_domain[0] = Has subdevice power domain
    power_domain[1] = Has device power domain
```

On the first pass through the for loop on line ~240, the ```else()``` case is hit and ```num_subdevice_power_domain``` is incremented.  On the next pass, the ```if()``` case is hit, ```num_device_power_domain``` is incremented, and the throw on line ~252 is skipped.  When we get here to line ~265, we have 1 ```device_power_domain``` and 1 ```subdevice_power_domain```.  Is this an error condition?

If you want to sync on this and try to iron out all these error conditions and consolidate the checks, let me know.